File: runtime-src/Monk/Runtime/Integer.hs
{-# LANGUAGE OverloadedStrings #-}

module Monk.Runtime.Integer (integerOperation, integerValue, parseNumber, wrap) where

import Data.Bits
import Data.ByteString (ByteString)
import Data.ByteString.Char8 qualified as B
import Data.Char (isAsciiLower, isAsciiUpper, isDigit, ord)

wrap :: Integer -> Integer
wrap n = (n + 2 ^ (63 :: Int)) `mod` 2 ^ (64 :: Int) - 2 ^ (63 :: Int)

parseNumber :: ByteString -> Either ByteString Integer
parseNumber input | B.null input = Right 0
parseNumber input = do
  let (sign, s) = case B.uncons input of Just ('-', r) -> (-1, r); Just ('+', r) -> (1, r); _ -> (1, input)
  (base, digits) <-
    if B.elem '#' s
      then case B.break (== '#') s of
        (b, r) | not (B.null b) && B.length b <= 2 && B.head b /= '0' && B.all isDigit b -> Right (foldl' (\n c -> n * 10 + toInteger (ord c - 48)) 0 (B.unpack b), B.tail r)
        _ -> Left "invalid-number"
      else
        if B.take 2 s `elem` ["0x", "0X"]
          then Right (16, B.drop 2 s)
          else if B.length s > 1 && B.head s == '0' then Right (8, B.tail s) else Right (10, s)
  if B.null digits || base < 2 || base > 64 then Left "invalid-number" else fmap (wrap . (sign *)) (foldl' (step base) (Right 0) (B.unpack digits))
  where
    step base acc c = do
      n <- acc
      let d
            | isDigit c = ord c - 48
            | isAsciiLower c = ord c - ord 'a' + 10
            | isAsciiUpper c = ord c - ord 'A' + if base <= 36 then 10 else 36
            | c == '@' = 62
            | c == '_' = 63
            | otherwise = 65
      if toInteger d >= base then Left "invalid-number" else Right (wrap (n * base + toInteger d))

integerValue :: ByteString -> [ByteString] -> Either ByteString Integer
integerValue "batch" args = batch [] args
  where
    batch [value] [] = Right value
    batch stack ("push" : spelling : rest) = do
      value <- parseNumber spelling
      batch (value : stack) rest
    batch (value : stack) (operation : rest) | operation `elem` ["pos", "neg", "not", "invert"] = do
      result <- integerValue operation [B.pack (show value)]
      batch (result : stack) rest
    batch (right : left : stack) (operation : rest) | operation `elem` ["add", "sub", "mul", "shl", "shr", "lt", "le", "gt", "ge", "eq", "ne", "and", "xor", "or"] = do
      result <- integerValue operation [B.pack (show left), B.pack (show right)]
      batch (result : stack) rest
    batch _ _ = Left "invalid-batch"
integerValue op args = do
  values <- traverse parseNumber args
  let bool p = if p then 1 else 0
  fmap wrap $ case (op, values) of
    ("read", [a]) -> Right a
    ("pos", [a]) -> Right a
    ("neg", [a]) -> Right (-a)
    ("not", [a]) -> Right (bool (a == 0))
    ("invert", [a]) -> Right (complement a)
    ("add", [a, b]) -> Right (a + b)
    ("sub", [a, b]) -> Right (a - b)
    ("mul", [a, b]) -> Right (a * b)
    ("div", [_, 0]) -> Left "division-by-zero"
    ("rem", [_, 0]) -> Left "division-by-zero"
    ("div", [a, b]) -> Right (a `quot` b)
    ("rem", [a, b]) -> Right (a `rem` b)
    ("pow", [a, b])
      | b >= 0 -> Right (power a b 1)
      | otherwise -> Left "negative-exponent"
    ("shl", [a, b]) -> Right (shiftL a (fromInteger (b .&. 63)))
    ("shr", [a, b]) -> Right (shiftR a (fromInteger (b .&. 63)))
    ("lt", [a, b]) -> Right (bool (a < b))
    ("le", [a, b]) -> Right (bool (a <= b))
    ("gt", [a, b]) -> Right (bool (a > b))
    ("ge", [a, b]) -> Right (bool (a >= b))
    ("eq", [a, b]) -> Right (bool (a == b))
    ("ne", [a, b]) -> Right (bool (a /= b))
    ("and", [a, b]) -> Right (a .&. b)
    ("xor", [a, b]) -> Right (xor a b)
    ("or", [a, b]) -> Right (a .|. b)
    ("logical-and", [a, b]) -> Right (bool (a /= 0 && b /= 0))
    ("logical-or", [a, b]) -> Right (bool (a /= 0 || b /= 0))
    _ -> Left "unknown-primitive"
  where
    power _ 0 acc = acc
    power a n acc = power (wrap (a * a)) (n `quot` 2) (if odd n then wrap (acc * a) else acc)

integerOperation :: [ByteString] -> Either ByteString ByteString
integerOperation ("batch" : args) = case integerValue "batch" args of
  Left reason -> Left reason
  Right value -> Right ("ok\n" <> B.pack (show value) <> "\n-\n")
integerOperation (op : args)
  | op `elem` unary && length args == 1 || op `elem` binary && length args == 2 = Right $ case integerValue op args of
      Left e -> "error\n-\n" <> e <> "\n"
      Right n -> "ok\n" <> B.pack (show n) <> "\n-\n"
  where
    unary = ["read", "pos", "neg", "not", "invert"]
    binary = ["add", "sub", "mul", "div", "rem", "pow", "shl", "shr", "lt", "le", "gt", "ge", "eq", "ne", "and", "xor", "or", "logical-and", "logical-or"]
integerOperation _ = Left "invalid integer operation or arity"

File: runtime/src/semantics/integer.rs
//! bash53-i64 arithmetic: all arithmetic is explicitly wrapping, including parsing.
// This bounded scalar remains private to arithmetic; neither wire frames nor
// FFI structures expose experimental pattern-type layout.
use crate::abi2::opcode::integer as opcode;

type ShiftBits = std::pat::pattern_type!(u32 is 0..=63);
struct ShiftCount(ShiftBits);
impl ShiftCount {
    fn masked(value: i64) -> Self {
        let bits = (value & 63) as u32;
        // SAFETY: masking proves the complete validity range 0..=63.
        Self(unsafe { std::mem::transmute::<u32, ShiftBits>(bits) })
    }
    fn get(self) -> u32 {
        // SAFETY: the pattern type has the u32 representation; all its values
        // are valid u32 values, and only masked() constructs this wrapper.
        unsafe { std::mem::transmute::<ShiftBits, u32>(self.0) }
    }
}

fn invalid_number<T>() -> Result<T, Vec<u8>> {
    Err(b"invalid-number".to_vec())
}

pub fn parse_number(input: &[u8]) -> Result<i64, Vec<u8>> {
    if input.is_empty() {
        return Ok(0);
    }
    let (negative, s) = match input[0] {
        b'-' => (true, &input[1..]),
        b'+' => (false, &input[1..]),
        _ => (false, input),
    };
    let (base, digits) = if let Some(at) = s.iter().position(|&c| c == b'#') {
        let prefix = &s[..at];
        if prefix.is_empty()
            || prefix.len() > 2
            || prefix[0] == b'0'
            || !prefix.iter().all(u8::is_ascii_digit)
        {
            return invalid_number();
        }
        (
            prefix
                .iter()
                .fold(0_u32, |n, c| n * 10 + u32::from(c - b'0')),
            &s[at + 1..],
        )
    } else if s.starts_with(b"0x") || s.starts_with(b"0X") {
        (16, &s[2..])
    } else if s.len() > 1 && s[0] == b'0' {
        (8, &s[1..])
    } else {
        (10, s)
    };
    if digits.is_empty() || !(2..=64).contains(&base) {
        return invalid_number();
    }
    let mut n = 0_i64;
    for &c in digits {
        let digit = match c {
            b'0'..=b'9' => u32::from(c - b'0'),
            b'a'..=b'z' => u32::from(c - b'a') + 10,
            b'A'..=b'Z' => u32::from(c - b'A') + if base <= 36 { 10 } else { 36 },
            b'@' => 62,
            b'_' => 63,
            _ => 65,
        };
        if digit >= base {
            return invalid_number();
        }
        n = n
            .wrapping_mul(i64::from(base))
            .wrapping_add(i64::from(digit));
    }
    Ok(if negative { n.wrapping_neg() } else { n })
}
fn unary(op: &[u8]) -> bool {
    matches!(
        op,
        opcode::READ | opcode::POS | opcode::NEG | opcode::NOT | opcode::INVERT
    )
}
fn binary(op: &[u8]) -> bool {
    matches!(
        op,
        opcode::ADD
            | opcode::SUB
            | opcode::MUL
            | opcode::DIV
            | opcode::REM
            | opcode::POW
            | opcode::SHL
            | opcode::SHR
            | opcode::LT
            | opcode::LE
            | opcode::GT
            | opcode::GE
            | opcode::EQ
            | opcode::NE
            | opcode::AND
            | opcode::XOR
            | opcode::OR
            | opcode::LOGICAL_AND
            | opcode::LOGICAL_OR
    )
}
fn evaluate(op: &[u8], args: &[i64]) -> Result<i64, Vec<u8>> {
    Ok(match (op, args) {
        (opcode::READ | opcode::POS, [a]) => *a,
        (opcode::NEG, [a]) => a.wrapping_neg(),
        (opcode::NOT, [a]) => i64::from(*a == 0),
        (opcode::INVERT, [a]) => !a,
        (opcode::ADD, [a, b]) => a.wrapping_add(*b),
        (opcode::SUB, [a, b]) => a.wrapping_sub(*b),
        (opcode::MUL, [a, b]) => a.wrapping_mul(*b),
        (opcode::DIV | opcode::REM, [_, 0]) => return Err(b"division-by-zero".to_vec()),
        (opcode::DIV, [a, b]) => a.wrapping_div(*b),
        (opcode::REM, [a, b]) => a.wrapping_rem(*b),
        (opcode::POW, [a, b]) => {
            if *b < 0 {
                return Err(b"negative-exponent".to_vec());
            }
            let (mut base, mut exponent, mut acc) = (*a, *b, 1_i64);
            while exponent != 0 {
                if exponent & 1 != 0 {
                    acc = acc.wrapping_mul(base);
                }
                base = base.wrapping_mul(base);
                exponent /= 2;
            }
            acc
        }
        (opcode::SHL, [a, b]) => a.wrapping_shl(ShiftCount::masked(*b).get()),
        (opcode::SHR, [a, b]) => a.wrapping_shr(ShiftCount::masked(*b).get()),
        (opcode::LT, [a, b]) => i64::from(a < b),
        (opcode::LE, [a, b]) => i64::from(a <= b),
        (opcode::GT, [a, b]) => i64::from(a > b),
        (opcode::GE, [a, b]) => i64::from(a >= b),
        (opcode::EQ, [a, b]) => i64::from(a == b),
        (opcode::NE, [a, b]) => i64::from(a != b),
        (opcode::AND, [a, b]) => a & b,
        (opcode::XOR, [a, b]) => a ^ b,
        (opcode::OR, [a, b]) => a | b,
        (opcode::LOGICAL_AND, [a, b]) => i64::from(*a != 0 && *b != 0),
        (opcode::LOGICAL_OR, [a, b]) => i64::from(*a != 0 || *b != 0),
        _ => return Err(b"unknown-primitive".to_vec()),
    })
}
pub fn integer_value(op: &[u8], args: &[Vec<u8>]) -> Result<i64, Vec<u8>> {
    if op != opcode::BATCH {
        return evaluate(
            op,
            &args
                .iter()
                .map(|a| parse_number(a))
                .collect::<Result<Vec<_>, _>>()?,
        );
    }
    let mut stack = Vec::new();
    let mut at = 0;
    while at < args.len() {
        let op = args[at].as_slice();
        at += 1;
        if op == opcode::PUSH && at < args.len() {
            stack.push(parse_number(&args[at])?);
            at += 1;
        } else if matches!(op, opcode::POS | opcode::NEG | opcode::NOT | opcode::INVERT)
            && !stack.is_empty()
        {
            let index = stack.len() - 1;
            stack[index] = evaluate(op, &stack[index..])?;
        } else if matches!(
            op,
            opcode::ADD
                | opcode::SUB
                | opcode::MUL
                | opcode::SHL
                | opcode::SHR
                | opcode::LT
                | opcode::LE
                | opcode::GT
                | opcode::GE
                | opcode::EQ
                | opcode::NE
                | opcode::AND
                | opcode::XOR
                | opcode::OR
        ) && stack.len() >= 2
        {
            let index = stack.len() - 2;
            let result = evaluate(op, &stack[index..])?;
            stack.truncate(index);
            stack.push(result);
        } else {
            return Err(b"invalid-batch".to_vec());
        }
    }
    if stack.len() == 1 {
        Ok(stack[0])
    } else {
        Err(b"invalid-batch".to_vec())
    }
}
pub fn integer_operation(frames: &[Vec<u8>]) -> Result<Vec<u8>, Vec<u8>> {
    let Some((op, args)) = frames.split_first() else {
        return Err(b"invalid integer operation or arity".to_vec());
    };
    if op == opcode::BATCH {
        return Ok(format!("ok\n{}\n-\n", integer_value(op, args)?).into_bytes());
    }
    if !(unary(op) && args.len() == 1 || binary(op) && args.len() == 2) {
        return Err(b"invalid integer operation or arity".to_vec());
    }
    Ok(match integer_value(op, args) {
        Ok(value) => format!("ok\n{value}\n-\n").into_bytes(),
        Err(reason) => [b"error\n-\n".as_slice(), &reason, b"\n"].concat(),
    })
}

#[cfg(test)]
mod shift_tests {
    use super::*;
    #[test]
    fn masked_shift_count_covers_boundary_values() {
        for input in [i64::MIN, -65, -64, -1, 0, 1, 63, 64, 65, i64::MAX] {
            assert_eq!(ShiftCount::masked(input).get(), (input & 63) as u32);
        }
    }
}
